server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java

```diff
@@ -117,7 +119,12 @@ public void getSegmentFiles(
                         assert directoryFiles.contains(file) == false : "Local store already contains the file " + file;
                         toDownloadSegmentNames.add(file);
                     }
-                    indexShard.getFileDownloader().download(remoteDirectory, storeDirectory, toDownloadSegmentNames);
+                    indexShard.getFileDownloader()
+                        .download(
+                            remoteDirectory,
+                            new ReplicationStatsDirectoryWrapper(storeDirectory, fileProgressTracker),
```

Comment:
So we're wrapping the `copyFrom` method on `storeDirectory`, then inside copyFrom we wrap the passed in `from` directory and instrument the `readBytes` method. Would it be simpler to just wrap the `from` directory here like the following?

```java
final Directory statsTrackingRemoteDirectory = new FilterDirectory(remoteDirectory) {
    @Override
    public IndexInput openInput(String name, IOContext context) throws IOException {
        final IndexInput input = in.openInput(name, context);
        return new IndexInput("StatsDirectoryWrapper(" + input.toString() + ")") {
            @Override
            public void close() throws IOException {
                input.close();
            }

            @Override
            public long getFilePointer() {
                throw new UnsupportedOperationException("only straight copies are supported");
            }

            @Override
            public void seek(long pos) throws IOException {
                throw new UnsupportedOperationException("seeks are not supported");
            }

            @Override
            public long length() {
                return input.length();
            }

            @Override
            public IndexInput slice(String sliceDescription, long offset, long length) throws IOException {
                throw new UnsupportedOperationException("slices are not supported");
            }

            @Override
            public byte readByte() throws IOException {
                throw new UnsupportedOperationException("use a buffer if you wanna perform well");
            }

            @Override
            public void readBytes(byte[] b, int offset, int len) throws IOException {
                // we rely on the fact that copyFrom uses a buffer
                input.readBytes(b, offset, len);
                fileProgressTracker.accept(dest, (long) len);
            }
        };
    }
};
indexShard.getFileDownloader().download(
    statsTrackingRemoteDirectory,
    storeDirectory,
    toDownloadSegmentNames);
```

Reply:
The added benefit with the current approach is we can also wrap `PrimaryShardReplicationSource` with the same class from `SegmentReplicationSource` and reuse it to move away from `MultiFileWriter`. It should be a quick change to split out the logic but I have limited this PR currently to just fix the bug.